system/core/database.php

```diff
@@ -599,10 +599,40 @@ function __construct($filename) {
 		function lock() {
 			if ($this->fd)
 				throw new LogicException('Cannot lock file ' . $this->filename . ', already locked');
-			$this->fd = fopen($this->filename, 'c+');
-			if ($this->fd === false)
-				throw new RuntimeException('Cannot lock file ' . $this->filename . ', open failed: ' . error_get_last()['message']);
-			flock($this->fd, LOCK_EX);
+
+			// Clear the last error, to prevent using an
+			// older error when something goes wrong without
+			// generating a PHP error (e.g. like read/write
+			// before PHP 7.4).
+			error_clear_last();
+
+			for ($attempt = 0; $attempt < 10; ++$attempt) {
+				$fd = fopen($this->filename, 'c+');
+				if ($fd === false)
+					throw new RuntimeException('Cannot lock file ' . $this->filename . ', open failed: ' . error_get_last()['message']);
+
+				if (!flock($fd, LOCK_EX)) {
+					fclose($fd);
+					throw new RuntimeException('Cannot lock file ' . $this->filename . ', lock failed: ' . error_get_last()['message']);
+				}
+
+				// If the locked file still exists
+				// (non-zero link count), we're done.
+				// Otherwise, another process has replaced the
+				// file and we still have the
+				// old (now replaced) file open, so
+				// retry opening and locking the new
+				// file. Note: This makes all locking
+				// fail when the file is hardlinked
+				// elsewhere.
+				if (fstat($fd)['nlink'] != 0) {
```

Comment:
Wat is het standpunt wat betreft type juggling?

Reply:
Wat bedoel je hier precies? Dat we eigenlijk `!==` zouden moeten gebruiken?

Reply:
ik wilde het aandragen als suggestie en was benieuwd naar jouw standpunt hierover. Voor mij "springt" `!=` er altijd uit in js en php. In c(++) onderdruk ik de neiging om mezelf er van te laten opkijken ;-)

Reply:
Ik programmeer meer in C/C++, waar dit niet zo'n issue is (en !== uberhaupt niet bestaat), dus ik ben er niet zo op bedacht :-) Ik geloof dat Laurens waar mogelijk ook stricte checks gebruikt, en hier lijkt het me ook eigenlijk wel een goed idee.

OTOH, stel dat er een keer `null` oid uit zou komen (kan nu volgens mij niet, maar wie weet is er nog een keer een platform waarop dat wel gebeurt), dan zou bij `!== 0` de if triggeren, terwijl er eigenlijk geen (of niet zeker) een geldig lock is. Dus eigenlijk juist beter als dan de *if* niet triggert (en als dat dan 10x hetzelfde is, krijg je een exception, wat wsl beter is dan geen lock terwijl je denkt dat je er wel 1 hebt). Ik heb er nu `> 0` van gemaakt, dat heeft hetzelfde effect, maar is misschien nog iets duidelijker (het is eigenlijk niet 0 die speciaal is, maar > 0).

Reply:
Kan ik me helemaal in vinden